src/JsonSchema/Constraints/TypeConstraint.php

```diff
@@ -249,14 +267,46 @@ protected function toNumber($value)
         if (is_numeric($value)) {
             return $value + 0; // cast to number
         }
+        if (is_bool($value) || is_null($value)) {
+            return (int) $value;
+        }
+        if ($this->getTypeCheck()->isArray($value) && count($value) === 1) {
+            reset($value);
+
+            return $this->toNumber(current($value));
+        }
 
         return $value;
     }
 
+    /**
+     * Converts a value to an integer
+     *
+     * @param mixed $value
+     *
+     * @return mixed
+     */
     protected function toInteger($value)
     {
-        if (is_numeric($value) && (int) $value == $value) {
-            return (int) $value; // cast to number
+        $numberValue = $this->toNumber($value);
```

Comment:
Basically the logic is to apply all the coercion rules for number, and then see if the result can be further coerced to an int. If it is, then return the cast int - if not, return the original value.